bpf/tests/bpf-prog-test-run/bpf-prog-test-run.go

```diff
@@ -0,0 +1,487 @@
+}
+
+func mapDeleteAll(m *ebpf.Map) error {
+	for {
+		key, err := m.NextKeyBytes(nil)
+		if err != nil {
+			return err
+		}
+		if key == nil {
+			break
+		}
+
+		err = m.Delete(key)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func testNop(spec *ebpf.Collection) error {
+	prog := spec.Programs["test_nop"]
+	if prog == nil {
+		return errors.New("did not find test_nop program")
+	}
+
+	packetIn := gopacket.NewSerializeBuffer()
+	opts := gopacket.SerializeOptions{}
+	gopacket.SerializeLayers(packetIn, opts,
+		&layers.Ethernet{
+			SrcMAC:       net.HardwareAddr{1, 0, 0, 3, 0, 10},
+			DstMAC:       net.HardwareAddr{1, 0, 0, 3, 0, 20},
+			EthernetType: layers.EthernetTypeIPv4,
+		},
+	)
+
+	bpfRet, bufOut, err := prog.Test(packetIn.Bytes())
+	if err != nil {
+		return fmt.Errorf("test run failed: %v", err)
+	}
+
+	if bpfRet != 0 { // CT_ACT_OK
+		return errors.New("unexpected return value")
+	}
+
+	if !reflect.DeepEqual(bufOut, packetIn.Bytes()) {
+		return errors.New("unexpected data modification")
+	}
+
+	return nil
+}
+
+func testMap(spec *ebpf.Collection) error {
+	prog := spec.Programs["test_map"]
+	if prog == nil {
+		return errors.New("did not find test_nop program")
+	}
+
+	bpfCtMap := spec.Maps["test_cilium_ct_tcp4_65535"]
+
+	mapBefore, err := ct4TcpMapFromBPF(bpfCtMap)
+	if err != nil {
+		return fmt.Errorf("ct4TcpMapFromBPF failed: %v", err)
+	}
+	if len(mapBefore) != 0 {
+		return errors.New("BPF CT map is not empty")
+	}
+
+	packetIn := gopacket.NewSerializeBuffer()
+	opts := gopacket.SerializeOptions{}
+	gopacket.SerializeLayers(packetIn, opts,
+		&layers.Ethernet{
+			SrcMAC:       net.HardwareAddr{1, 0, 0, 3, 0, 10},
+			DstMAC:       net.HardwareAddr{1, 0, 0, 3, 0, 20},
+			EthernetType: layers.EthernetTypeIPv4,
+		},
+	)
+
+	bpfRet, bufOut, err := prog.Test(packetIn.Bytes())
+	if err != nil {
+		return fmt.Errorf("test run failed: %v", err)
+	}
+
+	if bpfRet != 0 { // CT_ACT_OK
+		return errors.New("unexpected return value")
+	}
+
+	if !reflect.DeepEqual(bufOut, packetIn.Bytes()) {
+		return errors.New("unexpected data modification")
+	}
+
+	keyExpected := ctmap.CtKey4Global{
+		TupleKey4Global: tuple.TupleKey4Global{
+			TupleKey4: tuple.TupleKey4{
+				SourceAddr: types.IPv4{1, 1, 1, 1},
+				DestAddr:   types.IPv4{1, 1, 1, 2},
+				SourcePort: byteorder.HostToNetwork(uint16(1001)).(uint16),
+				DestPort:   byteorder.HostToNetwork(uint16(1002)).(uint16),
+				NextHeader: 0,
+				Flags:      0,
+			},
+		},
+	}
+	valExecpted := ctmap.CtEntry{
+		TxPackets: 1000,
+		RxPackets: 1000,
+	}
+	mapExpected := ct4GlobalMap{
+		keyExpected: valExecpted,
+	}
+
+	mapAfter, err := ct4TcpMapFromBPF(bpfCtMap)
+	if err != nil {
+		return fmt.Errorf("ct4TcpMapFromBPF failed: %v", err)
+	}
+
+	if !reflect.DeepEqual(mapAfter, mapExpected) {
+		return errors.New("resulting BPF CT map is not equal to expected")
+	}
+
+	mapDeleteAll(bpfCtMap)
+	return nil
+}
+
+func patchElf(fname string, fnameOut string) error {
```

Comment:
Definitely is, thanks! Added a comment.

> If we switch to use cilium/ebpf as our loader, it will likely not be necessary anymore

I thought that this (i.e., not processing al the fields of `bpf_elf_map`) was something that should be fixed on cilium/ebpf. For the current tests, it does not matter because we  don't care about these fields but for the loader I think we do. Nothing to do with this PR, just thought I'd mention it.